File: app/src/main/kotlin/ai/onspace/notes/data/storage/StorageService.kt
package ai.onspace.notes.data.storage

import android.content.Context
import android.content.SharedPreferences
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import ai.onspace.notes.data.encryption.EncryptionService
import ai.onspace.notes.data.model.Note
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class StorageService(context: Context) {

    private val prefs: SharedPreferences =
        context.getSharedPreferences("notes_encrypted_prefs", Context.MODE_PRIVATE)
    private val gson = Gson()
    private val NOTES_KEY = "notes_encrypted"

    suspend fun loadNotes(): List<Note> = withContext(Dispatchers.IO) {
        try {
            val encryptedData = prefs.getString(NOTES_KEY, null) ?: return@withContext emptyList()
            val decrypted = EncryptionService.decrypt(encryptedData)
            val type = object : TypeToken<List<Note>>() {}.type
            gson.fromJson(decrypted, type) ?: emptyList()
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun saveNotes(notes: List<Note>) = withContext(Dispatchers.IO) {
        val json = gson.toJson(notes)
        val encrypted = EncryptionService.encrypt(json)
        prefs.edit().putString(NOTES_KEY, encrypted).apply()
    }

    suspend fun createNote(title: String, content: String): Note = withContext(Dispatchers.IO) {
        val now = System.currentTimeMillis()
        val note = Note(
            id = now.toString(),
            title = title.ifBlank { "Без названия" },
            content = content,
            createdAt = now,
            updatedAt = now
        )
        val notes = loadNotes().toMutableList()
        notes.add(0, note)
        saveNotes(notes)
        note
    }

    suspend fun updateNote(id: String, title: String, content: String) = withContext(Dispatchers.IO) {
        val notes = loadNotes().toMutableList()
        val index = notes.indexOfFirst { it.id == id }
        if (index != -1) {
            notes[index] = notes[index].copy(
                title = title.ifBlank { "Без названия" },
                content = content,
                updatedAt = System.currentTimeMillis()
            )
            saveNotes(notes)
        }
    }

    suspend fun deleteNote(id: String) = withContext(Dispatchers.IO) {
        val notes = loadNotes().filter { it.id != id }
        saveNotes(notes)
    }
}

File: services/storage.service.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { encryptText, decryptText } from './encryption.service';

const NOTES_STORAGE_KEY = '@notes_encrypted';

export interface Note {
  id: string;
  title: string;
  content: string;
  createdAt: number;
  updatedAt: number;
}

/**
 * Загрузить все заметки из хранилища
 */
export async function loadNotes(): Promise<Note[]> {
  try {
    const encryptedData = await AsyncStorage.getItem(NOTES_STORAGE_KEY);
    if (!encryptedData) {
      return [];
    }
    
    const decryptedData = await decryptText(encryptedData);
    return JSON.parse(decryptedData);
  } catch (error) {
    console.error('Ошибка загрузки заметок:', error);
    return [];
  }
}

/**
 * Сохранить все заметки в хранилище
 */
export async function saveNotes(notes: Note[]): Promise<void> {
  try {
    const jsonData = JSON.stringify(notes);
    const encryptedData = await encryptText(jsonData);
    await AsyncStorage.setItem(NOTES_STORAGE_KEY, encryptedData);
  } catch (error) {
    console.error('Ошибка сохранения заметок:', error);
    throw error;
  }
}

/**
 * Создать новую заметку
 */
export async function createNote(title: string, content: string): Promise<Note> {
  const note: Note = {
    id: Date.now().toString(),
    title: title || 'Без названия',
    content,
    createdAt: Date.now(),
    updatedAt: Date.now(),
  };
  
  const notes = await loadNotes();
  notes.unshift(note);
  await saveNotes(notes);
  
  return note;
}

/**
 * Обновить существующую заметку
 */
export async function updateNote(id: string, title: string, content: string): Promise<void> {
  const notes = await loadNotes();
  const index = notes.findIndex(n => n.id === id);
  
  if (index !== -1) {
    notes[index] = {
      ...notes[index],
      title: title || 'Без названия',
      content,
      updatedAt: Date.now(),
    };
    await saveNotes(notes);
  }
}

/**
 * Удалить заметку
 */
export async function deleteNote(id: string): Promise<void> {
  const notes = await loadNotes();
  const filtered = notes.filter(n => n.id !== id);
  await saveNotes(filtered);
}
